File: playwright-tests/tests/posts.spec.ts
import { test, expect } from "@playwright/test";
import { DatabaseHelper } from "./helpers/database-helper";

test.describe("Posts Management", () => {
  test.beforeEach(async ({ page }) => {
    await page.goto("/posts");
  });

  test.describe("Creating a new post", () => {
    test("allows user to create a published post", async ({ page }) => {
      await page.click("text=New post");

      await page.fill('input[name="post[title]"]', "Playwright Test Post");
      await page.fill(
        'textarea[name="post[content]"]',
        "This post was created using Playwright",
      );

      await page.check('input[type="checkbox"][name="post[published]"]');
      await page.click('input[type="submit"]');
      await expect(
        page.locator("text=Post was successfully created"),
      ).toBeVisible();
      await expect(page.locator("text=Playwright Test Post")).toBeVisible();
    });
  });

  test.describe("Viewing posts", () => {
    test("shows only published posts on index", async ({ page }) => {
      await DatabaseHelper.createPost(
        page,
        "Published via Playwright",
        "This is a published post for testing",
        true,
      );

      await DatabaseHelper.createPost(
        page,
        "Draft via Playwright",
        "This is a draft post for testing",
        false,
      );

      await page.goto("/posts");
      await expect(page.locator("text=Published via Playwright")).toBeVisible();
      await expect(page.locator("text=Draft via Playwright")).not.toBeVisible();
    });
  });
});

File: spec/features/posts_spec.rb
require 'rails_helper'

RSpec.describe 'Posts Management', type: :feature do
  describe 'Creating a new post' do
    it 'allows user to create a published post' do
      visit posts_path
      click_link 'New post'

      fill_in 'Title', with: 'RSpec Test Post'
      fill_in 'Content', with: 'This post was created using RSpec'
      check 'Published'

      click_button 'Create Post'

      expect(page).to have_content('Post was successfully created')
      expect(page).to have_content('RSpec Test Post')
    end
  end

  describe 'Viewing posts' do
    let!(:published_post) { create(:post, title: 'Published via RSpec') }
    let!(:draft_post) { create(:post, :draft, title: 'Draft via RSpec') }

    it 'shows only published posts on index' do
      visit posts_path

      expect(page).to have_content('Published via RSpec')
      expect(page).not_to have_content('Draft via RSpec')
    end
  end
end
